packages/search/src/autodetectVisitor.ts

```diff
@@ -0,0 +1,254 @@
+import {transform} from "@babel/core";
+import pluginSyntaxTypescript from "@babel/plugin-syntax-typescript";
+import pluginTransformReactJsx from "@babel/plugin-transform-react-jsx";
+import {NodePath} from "@babel/traverse";
+import * as t from "@babel/types";
+import * as path from "path";
+import {format} from "./format";
+import {Declarations, File, FileError, FileExport, VisitorState} from "./types";
+import {
+  getExportNames,
+  getSourceForClassDecl,
+  getSourceForDefaultExport,
+  getSourceForFunctionDecl,
+  getSourceForVariableDecl,
+} from "./utils/visitor";
+
+const isReactElement = (node: t.Node) => {
+  if (t.isCallExpression(node) && t.isMemberExpression(node.callee)) {
+    const object: t.Expression = node.callee.object;
+    return (
+      t.isIdentifier(object) &&
+      object.name === "React" &&
+      node.callee.property.name === "createElement"
+    );
+  }
+  return false;
+};
+
+const returnsReactElement = (node: t.Node): boolean => {
+  if (isReactElement(node)) {
+    return true;
+  } else if (t.isBlockStatement(node)) {
+    const lastStatement = node.body[node.body.length - 1];
+    if (t.isReturnStatement(lastStatement) && lastStatement.argument) {
+      return isReactElement(lastStatement.argument);
+    }
+  }
+  return false;
+};
+
+const declaresReactElement = (
+  node: t.VariableDeclaration | t.FunctionDeclaration | t.ClassDeclaration,
+) => {
+  if (t.isVariableDeclaration(node)) {
+    const declarator: t.VariableDeclarator = node.declarations[0];
+    const name: string = t.isIdentifier(declarator.id)
+      ? declarator.id.name
+      : "";
+    const target: t.Expression | null = declarator.init;
+    const elem = t.isArrowFunctionExpression(target) ? target.body : target;
+    return name && target && elem && returnsReactElement(elem);
+  } else if (t.isFunctionDeclaration(node)) {
+    return returnsReactElement(node.body);
+  } else if (
+    t.isClassDeclaration(node) &&
+    t.isMemberExpression(node.superClass)
+  ) {
+    return (
+      t.isIdentifier(node.superClass.object) &&
+      node.superClass.object.name === "React" &&
+      node.superClass.property.name === "Component"
+    );
+  }
+  return false;
+};
+
+const declarationDetection = (declarations: Declarations) => () => ({
+  visitor: {
+    VariableDeclaration(nodePath: NodePath<t.VariableDeclaration>) {
+      const declarator: t.VariableDeclarator = nodePath.node.declarations[0];
+      const name = t.isIdentifier(declarator.id) && declarator.id.name;
+      if (name) {
+        declarations[name] = format(
```

Comment:
The source gets saved here because it gets transformed to es5 and we want the source to include the JSX?

Reply:
Yes exactly, I assumed we would prefer to display the JSX instead of React.createElement etc. There are probably other ways around this like transpiling it, but this seemed like a relatively simple solution.

The other reason to save the source (although it wouldn't have to be here) is so that you can get it even when the export is separate from the declaration, eg:
const Comp = <div />
// bla bla bla
export default Comp;

Reply:
Right, makes sense 👍